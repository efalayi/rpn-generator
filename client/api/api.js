import 'whatwg-fetch'

const API_BASE_URL = `${process.env.API_URL}/api/v1`
console.log('API_URL: ', process.env.API_URL)

function handleResponse(response) {
  return response.json()
}

function handleError(error) {
  return error
}

const getPhoneNumbers = async () => {
  try {
    const response = await fetch(`${API_BASE_URL}/contacts/phone-numbers`, {
      credentials: 'same-origin',
    })
    return handleResponse(response)
  } catch (error) {
    return handleError(error)
  }
}

const generateNewPhoneNumbers = async () => {
  try {
    const response = await fetch(`${API_BASE_URL}/contacts/generate`, {
      credentials: 'same-origin',
    })
    return handleResponse(response)
  } catch (error) {
    return handleError(error)
  }
}

export default {
  generateNewPhoneNumbers,
  getPhoneNumbers
}
