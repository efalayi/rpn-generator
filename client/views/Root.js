import React from 'react'

const Root = () => (
  <div>
    <h1>This is the Root of the app</h1>
  </div>
)

export default Root
