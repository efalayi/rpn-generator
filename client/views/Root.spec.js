import React from 'react'
import ReactTestRendrer from 'react-test-renderer'
import Root from './Root'
import api from '../api/api'
import NumberHelper from '../helpers/numberHelper'

jest.mock('../api/api')

const props = {
  totalCount: 3,
  phoneNumbers: ['0867647905', '0654421897', '0436789564'],
  controlPanelActions: [
    {
      name: 'Sort ASC',
      handler: jest.fn()
    },
    {
      name: 'Sort DSC',
      handler: jest.fn()
    },
    {
      name: 'Find Max',
      handler: jest.fn()
    },
    {
      name: 'Find Min',
      handler: jest.fn()
    }
  ],
  resultPanelValues: [
    {
      name: 'Max',
      value: '0867647905'
    },
    {
      name: 'Min',
      value: '0436789564'
    }
  ]
}

describe('#Root', () => {
  describe('with undefined phone numbers', () => {
    api.getPhoneNumbers = jest.fn(() => ({
      error: new Error()
    }))
    const componentWrapper = ReactTestRendrer.create(<Root />)
    const componentInstance = componentWrapper.getInstance()
    it('should have a defined state', () => {
      expect(componentInstance.state).toHaveProperty('phoneNumbers', [])
      expect(componentInstance.state).toHaveProperty('maxNumber', '')
      expect(componentInstance.state).toHaveProperty('minNumber', '')
    })
  })
  describe('with defined phone numbers', () => {
    api.getPhoneNumbers = jest.fn(() => ({
      phoneNumbers: props.phoneNumbers
    }))
    const componentWrapper = ReactTestRendrer.create(<Root />)
    const componentTree = componentWrapper.toJSON()
    it('should render without crashing', () => {
      const componentInstance = componentWrapper.getInstance()
      expect(componentTree).toMatchSnapshot()
      expect(componentInstance.state).toHaveProperty('phoneNumbers', props.phoneNumbers)
      expect(componentInstance.state).toHaveProperty('maxNumber', '')
      expect(componentInstance.state).toHaveProperty('minNumber', '')
    })

    it('should update maxNumber in state when findMax is clicked', () => {
      const componentRoot = componentWrapper.root
      const findMaxButton = componentRoot.findByProps({
        className: 'find-max'
      })
      findMaxButton.props.handleClick()
      const componentState = componentWrapper.getInstance().state
      expect(componentState.maxNumber).toEqual('0867647905')
    })

    it('should update minNumber in state when findMin is clicked', () => {
      const componentRoot = componentWrapper.root
      const findMinButton = componentRoot.findByProps({
        className: 'find-min'
      })
      findMinButton.props.handleClick()
      const componentState = componentWrapper.getInstance().state
      expect(componentState.minNumber).toEqual('0436789564')
    })

    it('should sort phone numbers in ascending order when Sort Ascending button is clicked', () => {
      const componentRoot = componentWrapper.root
      const phoneNumbersInAscendingOrder = NumberHelper.sortPhoneNumbers(props.phoneNumbers, 'asc')
      const sortAscButton = componentRoot.findByProps({
        className: 'sort-ascending'
      })
      sortAscButton.props.handleClick()
      const componentState = componentWrapper.getInstance().state
      expect(componentState.phoneNumbers).toEqual(phoneNumbersInAscendingOrder)
    })

    it('should sort phone numbers in descending order when Sort Descending button is clicked', () => {
      const componentRoot = componentWrapper.root
      const phoneNumbersInDescendingOrder = NumberHelper.sortPhoneNumbers(props.phoneNumbers, 'dsc')
      const sortAscButton = componentRoot.findByProps({
        className: 'sort-descending'
      })
      sortAscButton.props.handleClick()
      const componentState = componentWrapper.getInstance().state
      expect(componentState.phoneNumbers).toEqual(phoneNumbersInDescendingOrder)
    })

    it('should generate new phoneNumbers when Generate New Numbers button is clicked', () => {
      api.generateNewPhoneNumbers = jest.fn(() => ({
        newPhoneNumbers: ['012345678932', '04562378194']
      }))
      const generateNewPhoneNumbersSpy = jest.spyOn(api, 'generateNewPhoneNumbers')
      const componentRoot = componentWrapper.root
      const generateNewNumbersButton = componentRoot.findByProps({
        className: 'solid generate-new-numbers'
      })
      generateNewNumbersButton.props.onClick()
      expect(generateNewPhoneNumbersSpy).toHaveBeenCalled()
      expect(generateNewPhoneNumbersSpy).toHaveBeenCalledTimes(1)
      generateNewPhoneNumbersSpy.mockRestore()
    })
  })
})
