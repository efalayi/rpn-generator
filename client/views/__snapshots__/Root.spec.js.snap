// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Root with defined phone numbers should render without crashing 1`] = `
<div
  className="container flex-container flex-column flex-center"
  id="root"
>
  <h3>
    Random Phone Number Generator
  </h3>
  <div
    className="app-bar"
  >
    <button
      className="solid generate-new-numbers"
      onClick={[Function]}
      type="button"
    >
      Generate New Numbers
    </button>
    <h6>
      Total Phone Numbers: 0
    </h6>
  </div>
  <div
    className="control-panel"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Sort Ascending
    </button>
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Sort Descending
    </button>
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Find Max
    </button>
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Find Min
    </button>
  </div>
  <div
    className="result-panel container"
  >
    <div
      className="one-half column"
    >
      <div
        className="flex-container flex-center"
      >
        <span>
          Maximum Number
        </span>
        <span>
          
        </span>
      </div>
    </div>
    <div
      className="one-half column"
    >
      <div
        className="flex-container flex-center"
      >
        <span>
          Minimum Number
        </span>
        <span>
          
        </span>
      </div>
    </div>
  </div>
  <table>
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Phone Number
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
